addons/binding/org.openhab.binding.nest/src/main/java/org/openhab/binding/nest/handler/NestBridgeHandler.java

```diff
@@ -163,144 +151,95 @@ private void refreshData() {
-                Thing thingSmokeDetector = getExistingDevice(smokeDetector.getDeviceId(), things);
-                if (thingSmokeDetector != null) {
-                    NestSmokeDetectorHandler handler = (NestSmokeDetectorHandler) thingSmokeDetector.getHandler();
-                    if (handler != null) {
-                        handler.updateSmokeDetector(smokeDetector);
+                    if (devices.getCameras() != null) {
+                        devices.getCameras().values().forEach(listener::onNewNestCameraData);
                     }
-                } else {
-                    for (NestDeviceAddedListener listener : listeners) {
-                        logger.debug("Found new smoke detector. {}", smokeDetector.getDeviceId());
-                        listener.onSmokeDetectorAdded(smokeDetector);
+                    if (devices.getSmokeDetectors() != null) {
+                        devices.getSmokeDetectors().values().forEach(listener::onNewNestSmokeDetectorData);
                     }
                 }
-            }
-        }
+        );
     }
 
-    private void compareStructure(Collection<Structure> structures) {
-        Bridge bridge = getThing();
-        List<Thing> things = bridge.getThings();
-
-        for (Structure struct : structures) {
-            Thing thingStructure = getExistingDevice(struct.getStructureId(), things);
-            if (thingStructure != null) {
-                NestStructureHandler handler = (NestStructureHandler) thingStructure.getHandler();
-                if (handler != null) {
-                    handler.updateStructure(struct);
-                }
-            } else {
-                for (NestDeviceAddedListener listener : listeners) {
-                    logger.debug("Found new structure {}", struct.getStructureId());
-                    listener.onStructureAdded(struct);
-                }
-            }
-        }
+    private void broadcastStructure(Collection<Structure> structures) {
+        structures.forEach(
+                structure -> listeners.forEach(l -> l.onNewNestStructureData(structure))
+        );
     }
 
-    private String buildQueryString(NestBridgeConfiguration config)
-            throws InterruptedException, TimeoutException, ExecutionException {
+    private String buildQueryString(NestBridgeConfiguration config) throws InvalidAccessTokenException {
         String stringAccessToken;
+
+        // FIXME should we ever expire access tokens?
         if (config.accessToken == null) {
             stringAccessToken = accessToken.getAccessToken();
+
             // Update the configuration and persist to the database.
             Configuration bridgeConfiguration = editConfiguration();
-            bridgeConfiguration.put("accessToken", stringAccessToken);
+            bridgeConfiguration.put(NestBridgeConfiguration.ACCESS_TOKEN, stringAccessToken);
             updateConfiguration(bridgeConfiguration);
```

Comment:
This didn't work for me.  Are you sure it works?  It wouldn't reload the values after I restarted openhab.

Reply:
I will check thanks for pointing it out.

Reply:
@sjka @kaikreuzer should I be able to update the bridge configuration in a persistent (reboot safe) mater with the above code? If not is it intended or a feature?

Reply:
Well, you can obviously only update the configuration if the bridge is stored in the database and not being defined through a things-file. But then yes: the updateConfiguration call will persist the information, so that it is available after a restart. For textually defined bridges, the access token will be available in configuration in memory, but will be lost after a restart. If the access token can be newly retrieved through a refreshToken, this wouldn't even be an issue. Otherwise,  the user himself will have to get hold of the accessToken and add it to his textual configuration.

Reply:
That sounds fair and in my tests so far it looked fine, so maybe the underlying issue which @pinkfish experienced was temporary.